### DISCUSSION TITLE
Rename packages and some assemblies for version 4.0

### DISCUSSION THREAD

**Discussion by CharliePoole** [nunit/nunit-console#1535] [Ideas]
~~Given that 4.0 will be breaking, I think it makes sense to start adding "4" to many of the package names:~~

#### NuGet

~~Two options here. I prefer "NUnit4" provided we can get exclusive use of that prefix from nuget.org.~~

* ~~NUnit4.ConsoleRunner OR NUnit.ConsoleRunner4~~
* ~~NUnit4.ConsoleRunner.NetCore OR NUnit.ConsoleRunner4.NetCore~~
* ~~NUnit4.Console OR NUnit.Console4~~
* ~~NUnit4.Engine OR NUnit.Engine4~~
* ~~NUnit4.Engine.Api OR NUnit.Engine4.Api~~

#### Chocolatey
* nunit-console-runner-v4

#### MSI
* ~~NUnit4.Console~~ NOTE: We no longer have an msi package.

#### Zip
* ~~NUnit4.Console.Zip~~ NOTE: We no longer have a zip package.

### Assemblies
* nunit4-console.exe
* ~~nunit4.engine.dll~~
* ~~nunit4.engine.core.dll~~
* ~~nunit4.engine.api.dll~~
* ~~nunit4-agent.exe~~
* ~~nunit4-agent-x86.exe~~

@nunit/engine-team This is a big deal for users, so I'd appreciate your comments, particularly if you see any problems!

**Comment by mikkelbu** [comment]
I've not thought much about this, so this is my initial thoughts/concerns:
* Given that the package will have a new name a lot of users might not update as they don't get any help from tooling that a V4 version of their V3 dependency is available. So they might never notice the new package
* Do we have any idea of how long v4 of the console/engine will live? If it will last as long as the V2 and V3, then I don't mind having the version number as part of the names, but if we expect a quicker releases of major versions, then it might be annoying
* Wrt. naming of the nuget packages, then I think the actually prefer the other naming - e,g, NUnit.ConsoleRunner4 - as this makes it more clear that it is the version 4 of the ConsoleRunner and not version 4 of NUnit. I think this is clear for us, but our versioning between the framework, engine/console, testadapter, analyzers often confuse the users as they expect these to be more in sync.

That being said I don't use NUnit at work (and only a little at home), so I don't think you should put too much value on my comments. I think some of the other users depends much more on NUnit than I do.

**Comment by OsirisTerje** [comment]
As @mikkelbu says, having the version number as part of the name, breaks the possibilities for upgrades, and in particular information about the presence of an upgrade.   
Using the major version number for that is more than enough information to the user, that it do have breaking changes. 

Having the version number in the name, as we did for NUnit3.Testadapter which is now in version 4, is something  I'm still regretting we did.  And others are confused also, that NUnit3 adapter can be used with NUnit version 4.  

And if the console starts with NUnit4, does it mean it doesn't work with NUnit version 3, or only with NUnit version 4, what about NUnit version 5, which may go out next year ?

Further, I don't believe others do this either, semantic versioning is a better way. 

Also, we do have exclusive use of NUnit, so just keeping that solves one more.

**Comment by manfred-brands** [comment]
I also think we should stick to a non-versioned name. Changing the major number in semver indicates an incompatible change. At work all our package references are of the form `[x.y.z, x+1.0.0)` to indicate a minimum version with feature and bugfix and allow any later version up to but not including the next major version. The move for that is done after manual investigation.

**Comment by CharliePoole** [comment]
OK, you've all pretty much convinced me and I have updated the original proposal. Except for a few points...

* Chocolatey installs only one package version at a time. They had side by side installation back in version 1, but dropped it in version2. The reasoning for that was that they do not have control over the actual installation programs run and there is no guarantee that those installers support side by side execution. I would lean toward extending the same logic to any package intended for global installation, such as the zip file. 

* With respect to assembly names, I think we need to use `nunit4-console.exe` to keep it separate from `nunit3-console.exe`. The distinction is needed for side by side operation as well. I believe that `nunit-console` would just lead to confusing users.

By the way, in writing this my  assumption was that major upgrades would continue to be as rare as they have been historically. Whether that's the plan for the future really depends on who takes this over after me.

**Comment by CharliePoole** [comment]
@nunit/engine-team Any further comments on this?

**Comment by manfred-brands** [comment]
> With respect to assembly names, I think we need to use nunit4-console.exe to keep it separate from nunit3-console.exe. The distinction is needed for side by side operation as well. I believe that nunit-console would just lead to confusing users.

Most users would just want to install the next version of a tool and don't expect to then still have the old one installed.
For those few that want to do comparisons, the .net install for side-by-side would be unzip the .zip files into two separate directories.

When I run a tool from the command line, I don't know what version I have, nor do I care.
I don't want to have to specify `dotnet8 build` and `dotnet9 build` updating CI scripts when a new version comes out.

**Comment by CharliePoole** [comment]
@manfred-brands In almost all upgrade cases, I'd agree with you. But...

* It seems like we have to change the name from nunit3-console.exe, for the very reasons you list. If including '3' in that name was a mistake, it's one we made more than a decade ago. Obviously, we could go back to the old (V2) name, `nunit-console` if that seems preferable, but I think some change is required.
* We can't really provide a seamless upgrade in this case, since V4 has very big breaking changes. Old scripts will not run as before, no matter what name we use. (This is only true of the "standard" runner, not the dotnet tool, by the way.)
* Having gone through the V2 to V3 change, where so many users took years to upgrade, I'd like to make this as easy as possible. Part of that is likely to be side by side operation, at least where multiple projects are involved.
* Thinking about it, I guess this release will need to be accompanied by documentation of whatever we consider the best way to make the transition, tailored for shops of differing sizes, with varying numbers of projects in development.

**Comment by stevenaw** [comment]
I had partly typed up a response  when @CharliePoole replied and has answered a few questions already about historical reasons :)

I agree about minimizing friction during upgrade. Having to change all scripts to target a new executable is one kind of friction too, but perhaps a lesser one. Is there an issue which tracks the discussion leading to including the version number previously? I'd be curious to understand the decision making process at the time

**Reply by stevenaw** [reply]
Interesting. I have found the original v4 name thread ( https://github.com/nunit/nunit-console/issues/1149) where it seemed I asked a similar question and had it answered. I'll review that thread again now

**Reply by stevenaw** [reply]
@CharliePoole it seemed like the discussion there touched on a few interesting points including the nuget/chocolatey id and whether it matched the exe name. It seemed like the discussion settled on a version agnostic exe name while still allowing it to "percolate" as an idea. Were there other considerations not discussed there which made us lean back towards including the version number again? 

@jnm2 's point about forwarding scripts combined with different install directories is an interesting idea too (https://github.com/nunit/nunit-console/issues/1149#issuecomment-1055609532)

**Comment by CharliePoole** [comment]
@stevenaw I hadn't looked at that old thread for a few years, unfortunately. IIRC I kept the 4 in the name to make it easier for me to work on two versions at the same time without getting overly confused. I think the time is right to settle on the names we will use going forward.

It seems like there's a preference to remove the major version number from the assembly name. If we do that, we end up with the following names used to execute each runner.

* `nunit-console.exe`, to run the standard runner
* `dotnet nunit`, to run the dotnet tool, aka .NET Core Runner, which has a different feature set.
* `nunit-netcore-console.exe`, as an alternative way to run the .NET Core runner, provided the installation directory is accessible. (This is the internal name of the assembly, which is packaged as part of the tool.)

This leaves only one name with a version as part of it, the name of the chocolatey package. I simply don't know any way to avoid this since Chocolatey doesn't support installation of multiple versions of a package.

**Comment by OsirisTerje** [comment]
> This leaves only one name with a version as part of it, the name of the chocolatey package. I simply don't know any way to avoid this since Chocolatey doesn't support installation of multiple versions of a package.

Suggest the name is suffixed with the version in a way that makes it obvious it IS the version.   E.g.  NUnit.Console-4.1

**Reply by CharliePoole** [reply]
Good point. It has to conform with the Chocolatey community naming requirements, which is similar to what's used for Debian, so perhaps `nunit-console-runner-4` or `nunit-console-runner-v4`. I wouldn't want to use the minor version number, since we don't expect to change the name for minor version releases.

**Comment by CharliePoole** [comment]
Thanks for all the input. Since so many conflicting ideas are included in the thread, I'll summarize the conclusion...

 * One package change: `nunit-console-runner` => `nunit-console-runner-v4` in order to run size by side under chocolatey. We'll also need to create separate aliases in the chocolatey package.
 * One executable change: `nunit3-console.exe` => `nunit-console.exe`

See issue #1149 for follow-up.
